### PR TITLE
refactor: config

### LINKED ISSUE #15: Decorrelate CLI config from runner config
## Dependencies

- blocked by #4

### PR DESCRIPTION
**Pull request by GregoryAlbouy**
⚠️ **NO MERGE** until this PR targets main. Currently targeting https://github.com/benchttp/engine/pull/48

## Description

This PR follows https://github.com/benchttp/engine/pull/48 (extraction of CLI app from this repo).

Here we expose `configparse` as an external package, so it can be used by the newly standalone CLI app.

Some improvements around Config came in the process (see commits for details)

## Linked issues

- Closes #15

### CODE REVIEW

File: runner/internal/config/config.go [hunk 1]
```diff
@@ -83,13 +84,34 @@ type Output struct {
...
+}
+
+// WithField returns a new Global with the input fields marked as set.
+func (cfg Global) WithFields(fields ...string) Global {
+	fieldsSet := cfg.fieldsSet
+	if fieldsSet == nil {
+		fieldsSet = set{}
+	}
+	fieldsSet.add(fields...)
+	cfg.fieldsSet = fieldsSet
```
Comment (moreirathomas) [inline runner/internal/config/config.go:113]: Question: is there a reason why we do not mutate directly `cfg.fieldsSet`?

```suggestion
	if cfg.fieldsSet == nil {
		cfg.fieldsSet = set{}
	}
	cfg.fieldsSet.add(fields...)
```

File: runner/internal/config/config.go [hunk 2]
```diff
@@ -83,13 +84,34 @@ type Output struct {
...
+
 // Global represents the global configuration of the runner.
 // It must be validated using Global.Validate before usage.
 type Global struct {
 	Request Request
 	Runner  Runner
 	Output  Output
 	Tests   []tests.Case
+
+	fieldsSet set
```
Comment (moreirathomas) [inline runner/internal/config/config.go:103]: Suggestion: disambiguate "set" as the underlying data structure from the actual meaning "field is set by user/something, and is not at its default value".

```suggestion
	fieldsManuallySet set
```
or
```suggestion
	fieldsUndefaulted set
```

File: runner/internal/config/config.go [hunk 3]
```diff
@@ -99,37 +121,44 @@ func (cfg Global) String() string {
...
+	c.fieldsSet = nil
+	return reflect.DeepEqual(cfg, c)
+}
+
 // Override returns a new Config based on cfg with overridden values from c.
 // Only fields specified in options are replaced. Accepted options are limited
 // to existing Fields, other values are silently ignored.
-func (cfg Global) Override(c Global, fields ...string) Global {
-	for _, field := range fields {
+func (cfg Global) Override(c Global) Global {
```
Comment (moreirathomas) [inline runner/internal/config/config.go:134]: Suggestion for doc:

```suggestion
// Override returns a new Config by overriding the values on prev with the values present on cfg.
// Values present in prev and not in cfg are copied.
// Only fields specified with cfg.WithFields are replaced. Accepted options are limited
// to existing Fields, other values are silently ignored.
func (cfg Global) Override(prev Global) Global {
```

**Review by moreirathomas — COMMENTED** [review]
Finally made it ! Great work 😄

I have some suggestions for more explicit (?) names.

**Review by moreirathomas — COMMENTED** [review]
After reading through benchttp/cli#1, isn't `Config.Output.Silence` deprecated in the scope of this pr and should be remove?

**Comment by GregoryAlbouy** [comment]
> After reading through [benchttp/cli#1](https://github.com/benchttp/cli/pull/1), isn't `Config.Output.Silence` deprecated in the scope of this pr and should be remove?

Already done, forgot to push 🙃

**Review by moreirathomas — APPROVED** [review]

**Pull request merged**